services/contrib/PrometheusScrapeAgent/prometheus_scrape/agent.py

```diff
@@ -0,0 +1,208 @@
+        self.vip.pubsub.subscribe(peer='pubsub',
+                                  prefix=topics.DRIVER_TOPIC_BASE,
+                                  callback=self._capture_device_data)
+
+    @Core.receiver("onstop")
+    def _stopping(self, sender, **kwargs):
+        pass
+
+    def scrape(self, env, data):
+        scrape_time = get_utc_seconds_from_epoch()
+        keys_to_delete = defaultdict(list)
+        if len(self._cache) > 0:
+            result = "# TYPE volttron_data gauge\n"
+            for device, device_topics in self._cache.iteritems():
+                device_tags = device.replace("-", "_").split('/')
+                for topic, value in device_topics.iteritems():
+                    if value[1] + self._cache_time > scrape_time:
+                        metric_props = re.split(self._tag_delimiter_re,
+                                                topic.lower())
+                        metric_tag_str = (
+                            "campus=\"{}\",building=\"{}\","
+                            "device=\"{}\",").format(*device_tags)
+                        for i, prop in enumerate(metric_props):
+                            metric_tag_str += "tag{}=\"{}\",".format(i, prop)
+                        result += "{}{{{}topic=\"{}\"}} {}\n".format(
+                            re.sub(" |/|-", "_", device), metric_tag_str,
+                            topic.replace(" ", "_"), value[0])
+                    else:
+                        try:
+                            keys_to_delete[device].append(topic)
+                        except Exception as e:
+                            _log.error("Could not delete stale topic")
+                            _log.exception(e)
+                        continue
+        else:
+            result = "#No Data to Scrape"
+        for device, delete_topics in keys_to_delete.iteritems():
+            for topic in delete_topics:
+                del self._cache[device][topic]
+
+        return result, [('Content-Type', 'text/plain')]
+
+    def _clean_compat(self, sender, topic, headers, message):
+        try:
+            # 2.0 agents compatability layer makes sender == pubsub.compat so
+            # we can do the proper thing when it is here
+            if sender == 'pubsub.compat':
+                data = compat.unpack_legacy_message(headers, message)
+            else:
+                data = message
+            return data
+        except ValueError as e:
+            _log.error("message for {topic} bad message string: "
+                       "{message_string}".format(topic=topic,
+                                                 message_string=message[0]))
+            raise e
+
+        except IndexError as e:
+            _log.error("message for {topic} missing message string".format(
+                topic=topic))
+            raise e
+
+    def _capture_log_data(self, peer, sender, bus, topic, headers, message):
```

Comment:
Sorry, I wasn't clear, it will be used in the future. I can take a crack at that now, but I think I will put a flag to disable or enable datalogger collection, as I'm not very interested in that data. I use external telemetry to collect system metrics from the devices the platforms are provisioned on.